### PR TITLE
types: time_point_to_string: harden against out of range timestamps

### PR DESCRIPTION
**Pull request by bhalevy**
The time point is multiplied by an adjustment factor of 1000
for boost::posix_time::time_duration::ticks_per_second() = 1000000
when calling boost::posix_time::milliseconds(count)
and that may lead to integer overflow as reported
by the UndefinedBehaviorSanitizer.

See https://github.com/scylladb/scylla/issues/10830#issuecomment-1158899187

This change checks for possible overflow in advance and
prints the raw counter value in this case, along with
an explanation.

Refs #10830

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

### CODE REVIEW

File: types.cc [hunk 1]
```diff
@@ -67,13 +67,22 @@ requires requires {
...
+        if (count >= std::numeric_limits<int64_t>::max() / adjustment_ratio || count <= std::numeric_limits<int64_t>::min() / adjustment_ratio) {
+            return fallback_format("out of range");
+        }
+    }
   try {
+    auto time = boost::posix_time::from_time_t(0) + boost::posix_time::milliseconds(count);
     return boost::posix_time::to_iso_extended_string(time);
   } catch (const std::exception& e) {
-    return format("{} {} ({})", count, units, e.what());
+    return fallback_format(e.what());
```
Comment (avikivity) [inline types.cc:85]: Let's use fmt instead of all this complexity. See https://fmt.dev/latest/syntax.html#chrono-specs.

Comment (bhalevy) [reply types.cc]: Done in v2

**Comment by bhalevy** [comment]
In v2 (7660548):
- used fmt library to format the iso compliant string representation.

**Comment by bhalevy** [comment]
In v3 (931a6a127):
- Fix compatibility issue with `boost::posix_time::to_iso_extended_string`

File: types.cc [hunk 2]
```diff
@@ -67,14 +70,19 @@ requires requires {
...
-    auto time = boost::posix_time::from_time_t(0) + boost::posix_time::milliseconds(count);
-    constexpr std::string_view units = "milliseconds";
-  try {
-    return boost::posix_time::to_iso_extended_string(time);
-  } catch (const std::exception& e) {
-    return format("{} {} ({})", count, units, e.what());
-  }
+    auto d = std::div(int64_t(tp.time_since_epoch().count()), int64_t(1000));
+    std::time_t seconds = d.quot;
+    std::tm tm = *gmtime(&seconds);
```
Comment (nyh) [inline types.cc:75]: Do you know if gmtime uses a per-thread variable or uses the same one for all threads? The manual doesn't claim it uses a per-thread variable.

So please change to use gmtime_r.

File: types.cc [hunk 3]
```diff
@@ -67,14 +70,19 @@ requires requires {
...
+    if (!d.rem) {
+        return fmt::format("{:%Y-%m-%dT%H:%M:%S}", tm);
+    }
+    // adjust seconds for time points earlier than posix epoch
+    if (d.quot < 0) {
```
Comment (nyh) [inline types.cc:80]: Maybe you meant to check if d.rem < 0?
For example, what if d.quot is exactly 0 and d.rem is negative?

**Review by nyh — CHANGES_REQUESTED** [review]

File: types.cc (hunk 3 above)
Comment (bhalevy) [reply types.cc:80]: good point

**Comment by bhalevy** [comment]
In v4 (87ee6c3722):
- use gmtime_r
- check gmtime_r for null return - indicating year overflow
  - return `"{count} milliseconds (out of range)"` in this case
- use `millis` variable instead of `d.rem`
- check for `millis < 0` rather than `d.quot < 0` as per https://github.com/scylladb/scylla/pull/10831#discussion_r901858367
  - and recalculate gmtime_r(&seconds, &rm) after adjustment (was missing)
- use automatic variable types

File: types.cc [hunk 4]
```diff
@@ -67,14 +70,26 @@ requires requires {
...
-  }
+    auto count = tp.time_since_epoch().count();
+    auto d = std::div(int64_t(count), int64_t(1000));
+    std::time_t seconds = d.quot;
+    std::tm tm;
+    if (!gmtime_r(&seconds, &tm)) {
+        return fmt::format("{} milliseconds (out of range)", count);
+    }
+    auto millis = d.rem;
+    if (!millis) {
```
Comment (nyh) [inline types.cc:81]: Why is this an interesting case? Isn't it easier just not to have this if() and print 0 milliseconds?

**Review by nyh — APPROVED** [review]
Looks good now. 

Since you obviously had to check a lot of different cases to write this code correctly (overflowing dates, time before the epoch, time just a few miliseconds before the epoch), it would have been good to do these checks in the form of a unit test, that could help the next time that somebody wants to rewrite this function yet again.

File: types.cc (hunk 4 above)
Comment (bhalevy) [reply types.cc:81]: There's a unit test that fails without that...

**Comment by bhalevy** [comment]
> Since you obviously had to check a lot of different cases to write this code correctly (overflowing dates, time before the epoch, time just a few miliseconds before the epoch), it would have been good to do these checks in the form of a unit test, that could help the next time that somebody wants to rewrite this function yet again.

Added "test: types: add test cases for timestamp_type to_string format"

**Comment by xemul** [comment]
@bhalevy , this failed on my version of libfmt:
```
test/boost/types_test.cc(297): fatal error: in "test_timestamp_string_conversions": critical check timestamp_type->to_string(timestamp_type->decompose(tp)) == "0-01-01T00:00:00" has failed [0000-01-01T00:00:00 != 0-01-01T00:00:00]
```
because `%Y` produced `0000` year, not `0` as expected by the test.

**Pull request closed without merging**